src/dmd/expressionsem.d

```diff
@@ -3605,21 +3605,22 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             if (cd && cd.classKind == ClassKind.cpp && exp.f && !exp.f.fbody)
             {
                 // if super is defined in C++, it sets the vtable pointer to the base class
-                // so we have to rewrite it, but still return 'this' from super() call:
-                // (auto tmp = super(), this.__vptr = __vtbl, tmp)
-                __gshared int superid = 0;
-                char[20] buf;
-                sprintf(buf.ptr, "__super%d", superid++);
-                auto tmp = copyToTemp(0, buf.ptr, result);
+                // so we have to restore it, but still return 'this' from super() call:
+                // (auto __vptrTmp = this.__vptr, auto __superTmp = super()), (this.__vptr = __vptrTmp, __superTmp)
                 Loc loc = exp.loc;
-                Expression tmpdecl = new DeclarationExp(loc, tmp);
 
-                auto dse = new DsymbolExp(loc, cd.vtblSymbol());
-                auto ase = new AddrExp(loc, dse);
-                auto pte = new DotIdExp(loc, new ThisExp(loc), Id.__vptr);
-                auto ate = new AssignExp(loc, pte, ase);
+                auto vptr = new DotIdExp(loc, new ThisExp(loc), Id.__vptr);
+                auto vptrTmpDecl = copyToTemp(0, "__vptrTmp", vptr);
+                auto declareVptrTmp = new DeclarationExp(loc, vptrTmpDecl);
 
-                Expression e = new CommaExp(loc, new CommaExp(loc, tmpdecl, ate), new VarExp(loc, tmp));
+                auto superTmpDecl = copyToTemp(0, "__superTmp", result);
+                auto declareSuperTmp = new DeclarationExp(loc, superTmpDecl);
+
+                auto declareTmps = new CommaExp(loc, declareVptrTmp, declareSuperTmp);
+
+                auto restoreVptr = new AssignExp(loc, vptr.syntaxCopy(), new VarExp(loc, vptrTmpDecl));
+
+                Expression e = new CommaExp(loc, declareTmps, new CommaExp(loc, restoreVptr, new VarExp(loc, superTmpDecl)));
```

Comment:
When using `Expression.combine(a, b, c, d)` here, creating `((a, b), (c, d))`, all hell breaks loose, requiring additional `expressionSemantic()` calls; `combine(lhs, rhs)` additionally sets the new `CommaExp.type` to the rhs type, which seems to make `expressionSemantic()` not descend to both children.